src/modules/expression/precedence.rs

```diff
@@ -0,0 +1,266 @@
+use heraclitus_compiler::prelude::*;
+use crate::modules::command::cmd::Command;
+use crate::modules::types::{parse_type, Typed};
+use crate::utils::ParserMetadata;
+use super::expr::{Expr, ExprType};
+use super::ternop::ternary::Ternary;
+use super::parentheses::Parentheses;
+use crate::modules::variable::get::VariableGet;
+use crate::modules::function::invocation::FunctionInvocation;
+use crate::modules::builtin::nameof::Nameof;
+use crate::{parse_non_operators, parse_operator};
+use super::literal::{
+    bool::Bool,
+    number::Number,
+    text::Text,
+    array::Array,
+    null::Null,
+    status::Status
+};
+use super::binop::{
+    add::Add,
+    sub::Sub,
+    mul::Mul,
+    div::Div,
+    modulo::Modulo,
+    range::Range,
+    and::And,
+    or::Or,
+    gt::Gt,
+    ge::Ge,
+    lt::Lt,
+    le::Le,
+    eq::Eq,
+    neq::Neq,
+    cast::Cast,
+    is::Is
+};
+use super::unop::{
+    not::Not,
+    neg::Neg
+};
+
+impl Expr {
+    /// This function will parse expressions in the precedence order.
+    /// All the operators are left associative by default, meaning
+    /// that they are parsed from left to right
+    pub fn parse_expression(&self, meta: &mut ParserMetadata) -> Result<Expr, Failure> {
+        self.parse_ternary(meta)
+    }
```

Comment:
I don't like that you implement Expr twice, in different files. If you want to separate the traits, then do that - create a trait here, but implement it in one file - `expr.rs` for Expr.

It also doesn't make sense, to have all those binops in separate files, where there is only transpilation and adding comments and nothing more. You do most of the heavy lifting here in one file, but leave small things in multiple separate files.

I assume you're parsing things here, because you enforce precedence of expressions by calling parsing from least important, to most important. This is bad design in my opinion. It requires us to keep in mind precedence of every operation and how the given operation does parsing, so that when we add a new expression, we need to change how other expressions are parsed.

Why not make a list of expressions in descending precedence, that we would iterate over and try to parse?

Reply:
A macro has been created to improve the readability of this part of compiler.

<img width="510" alt="image" src="https://github.com/user-attachments/assets/6ebc572b-3e1c-4313-b0e6-1b5d3a3a90b6">